test/integration/src/test/java/io/pravega/test/integration/ReadWriteTest.java

```diff
@@ -0,0 +1,252 @@
+        String readerGroupName = "testMultiReaderWriterReaderGroup";
+        //20  readers -> 20 stream segments ( to have max read parallelism)
+        ScalingPolicy scalingPolicy = ScalingPolicy.fixed(20);
+        StreamConfiguration config = StreamConfiguration.builder().scope(scope)
+                .streamName(STREAM_NAME).scalingPolicy(scalingPolicy).build();
+        StreamManager streamManager = new StreamManagerImpl(controller);
+        ClientFactory clientFactory = new ClientFactoryImpl(scope, controller);
+
+        //create a scope
+        Boolean createScopeStatus = streamManager.createScope(scope);
+        log.info("Create scope status {}", createScopeStatus);
+        //create a stream
+        Boolean createStreamStatus = streamManager.createStream(scope, STREAM_NAME, config);
+        log.info("Create stream status {}", createStreamStatus);
+
+        eventsReadFromPravega = new ConcurrentLinkedQueue<>();
+        stopReadFlag = new AtomicBoolean(false);
+        eventData = new AtomicLong(); //data used by each of the writers.
+        eventReadCount = new AtomicLong(); // used by readers to maintain a count of events.
+
+        //start writing events to the stream
+        log.info("Creating {} writers", NUM_WRITERS);
+        List<CompletableFuture<Void>> writerList = new ArrayList<>();
+        for (i = 0; i < NUM_WRITERS; i++) {
+            log.info("Starting writer{}", i);
+            writerList.add(startNewWriter(eventData, clientFactory));
+        }
+
+        //create a reader group
+        log.info("Creating Reader group : {}", readerGroupName);
+        ReaderGroupManager readerGroupManager = new ReaderGroupManagerImpl(scope, controller, clientFactory);
+        readerGroupManager.createReaderGroup(readerGroupName, ReaderGroupConfig.builder().startingTime(0).build(),
+                Collections.singleton(STREAM_NAME));
+        log.info("Reader group name {} ", readerGroupManager.getReaderGroup(readerGroupName).getGroupName());
+        log.info("Reader group scope {}", readerGroupManager.getReaderGroup(readerGroupName).getScope());
+
+        //create readers
+        log.info("Creating {} readers", NUM_READERS);
+        List<CompletableFuture<Void>> readerList = new ArrayList<>();
+        String readerName = "reader" + new Random().nextInt(Integer.MAX_VALUE);
+        //start reading events
+        for (i = 0; i < NUM_READERS; i++) {
+            log.info("Starting reader{}", i);
+            readerList.add(startNewReader(readerName + i, clientFactory, readerGroupName,
+                    eventsReadFromPravega, eventData, eventReadCount, stopReadFlag));
+        }
+        log.info("Online readers {}", readerGroupManager.getReaderGroup(readerGroupName).getOnlineReaders());
+
+        //wait for writers completion
+        FutureHelpers.allOf(writerList);
+        // wait for reads = writes
+        while (TOTAL_NUM_EVENTS != eventsReadFromPravega.size()) {
+            Thread.sleep(5);
```

Comment:
This sleep here can be replaced with thread wait/notify mechanism. wait() on an object here. Any reader which hits the condition(TOTAL_NUM_EVENTS != eventsReadFromPravega.size()) can notify() the object